engine/crates/engine-v2/src/sources/graphql/federation.rs

```diff
@@ -151,6 +145,59 @@ impl FederationEntityResolver {
     }
 }
 
+async fn cache_fetches<'ctx, R: Runtime>(
+    ctx: ExecutionContext<'ctx, R>,
+    endpoint: schema::SchemaWalker<'_, GraphqlEndpointId>,
+    headers: &http::HeaderMap,
+    representations: Vec<Box<RawValue>>,
+    ingester: &mut EntityIngester<'ctx, R>,
+) -> CacheFetchOutcome {
+    let header_result = serde_json::to_vec(
+        &headers
+            .iter()
+            .map(|(name, value)| (name.as_str(), value.as_bytes()))
+            .collect::<Vec<_>>(),
+    )
+    .inspect_err(|err| tracing::warn!("Couldn't encode headers as JSON for cache key: {err}"));
+
+    let Ok(headers_json) = header_result else {
+        return CacheFetchOutcome::Other {
+            filtered_representations: representations,
+        };
+    };
+
+    let fetches = representations
+        .iter()
+        .map(|repr| cache_fetch(ctx, endpoint.subgraph_name(), &headers_json, repr));
+
+    let cache_entries = join_all(fetches).await;
+    let fully_cached = !cache_entries.iter().any(CacheEntry::is_miss);
+
+    ingester.cache_entries = Some(cache_entries);
```

Comment:
Wouldn't mutate the ingester here but return the `cache_entires` in the `CacheFetchOutcome`.